[namedtensor] Add named_guard to native_functions options

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21373 [namedtensor] Add named_guard to native_functions options**
* #21341 [namedtensor] Add unique_ptr<NamedTensorMeta> field to TensorImpl
* #21281 [namedtensor] Implement at::match(Dimname, Dimname) and at::unify(Dimname, Dimname)
* #21280 [namedtensor] Added at::Dimname

If named_guard=True (default), if any of the functions' inputs are named
tensors with non-None names, then the function errors out. The plan is
to incrementally write named inference rules and setting
named_guard=False once a function supports named tensors.

Test Plan:

Read generated code. For example:
```
Tensor & CPUType::_cat_out(Tensor & out, TensorList tensors, int64_t dim) const {
    if (out.is_named() || at::has_names(tensors)) {
        AT_ERROR("_cat_out: no named inference rule implemented.");
    }
    const OptionalDeviceGuard device_guard(device_of(out));
    return at::native::legacy::cpu::_th_cat_out(/* native_actuals */ out, tensors, dim);
}
```

gh-metadata: pytorch pytorch 21373 gh/zou3519/42/head

Differential Revision: [D15717925](https://our.internmc.facebook.com/intern/diff/D15717925)

zou3519 committed:
[namedtensor] Add named_guard to native_functions options

If named_guard=True (default), if any of the functions' inputs are named
tensors with non-None names, then the function errors out. The plan is
to incrementally write named inference rules and setting
named_guard=False once a function supports named tensors.

Test Plan:

Read generated code. For example:
```
Tensor & CPUType::_cat_out(Tensor & out, TensorList tensors, int64_t dim) const {
    if (out.is_named() || at::has_names(tensors)) {
        AT_ERROR("_cat_out: no named inference rule implemented.");
    }
    const OptionalDeviceGuard device_guard(device_of(out));
    return at::native::legacy::cpu::_th_cat_out(/* native_actuals */ out, tensors, dim);
}
```

zou3519 committed:
Update on "[namedtensor] Add named_guard to native_functions options"

[namedtensor] Add named_guard to native_functions options

If named_guard=True (default), if any of the functions' inputs are named
tensors with non-None names, then the function errors out. The plan is
to incrementally write named inference rules and setting
named_guard=False once a function supports named tensors.

Test Plan:

Read generated code. For example:
```
Tensor & CPUType::_cat_out(Tensor & out, TensorList tensors, int64_t dim) const {
    if (out.is_named() || at::has_names(tensors)) {
        AT_ERROR("_cat_out: no named inference rule implemented.");
    }
    const OptionalDeviceGuard device_guard(device_of(out));
    return at::native::legacy::cpu::_th_cat_out(/* native_actuals */ out, tensors, dim);
}
```

gh-metadata: pytorch pytorch 21373 gh/zou3519/42/head

gchanan commented:
@nairbv can you take a look?

zou3519 committed:
Update on "[namedtensor] Add named_guard to native_functions options"

[namedtensor] Add named_guard to native_functions options

If named_guard=True (default), if any of the functions' inputs are named
tensors with non-None names, then the function errors out. The plan is
to incrementally write named inference rules and setting
named_guard=False once a function supports named tensors.

Test Plan:

Read generated code. For example:
```
Tensor & CPUType::_cat_out(Tensor & out, TensorList tensors, int64_t dim) const {
    if (out.is_named() || at::has_names(tensors)) {
        AT_ERROR("_cat_out: no named inference rule implemented.");
    }
    const OptionalDeviceGuard device_guard(device_of(out));
    return at::native::legacy::cpu::_th_cat_out(/* native_actuals */ out, tensors, dim);
}
```

gh-metadata: pytorch pytorch 21373 gh/zou3519/42/head

zou3519 committed:
Update on "[namedtensor] Add named_guard to native_functions options"

[namedtensor] Add named_guard to native_functions options

If named_guard=True (default), if any of the functions' inputs are named
tensors with non-None names, then the function errors out. The plan is
to incrementally write named inference rules and setting
named_guard=False once a function supports named tensors.

Test Plan:

Read generated code. For example:
```
Tensor & CPUType::_cat_out(Tensor & out, TensorList tensors, int64_t dim) const {
    if (out.is_named() || at::has_names(tensors)) {
        AT_ERROR("_cat_out: no named inference rule implemented.");
    }
    const OptionalDeviceGuard device_guard(device_of(out));
    return at::native::legacy::cpu::_th_cat_out(/* native_actuals */ out, tensors, dim);
}
```

gh-metadata: pytorch pytorch 21373 gh/zou3519/42/head

zou3519 committed:
Update on "[namedtensor] Add named_guard to native_functions options"

[namedtensor] Add named_guard to native_functions options

If named_guard=True (default), if any of the functions' inputs are named
tensors with non-None names, then the function errors out. The plan is
to incrementally write named inference rules and setting
named_guard=False once a function supports named tensors.

Test Plan:

Read generated code. For example:
```
Tensor & CPUType::_cat_out(Tensor & out, TensorList tensors, int64_t dim) const {
    if (out.is_named() || at::has_names(tensors)) {
        AT_ERROR("_cat_out: no named inference rule implemented.");
    }
    const OptionalDeviceGuard device_guard(device_of(out));
    return at::native::legacy::cpu::_th_cat_out(/* native_actuals */ out, tensors, dim);
}
```

gh-metadata: pytorch pytorch 21373 gh/zou3519/42/head

nairbv commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -594,6 +607,20 @@ def device_guard(option, dispatch_options, dispatch_tensor):
 
 
+def named_guard(option, tensors, tensorlists):
+    if not option.get('named_guard', True) or (len(tensors) + len(tensorlists) == 0):
```
I think we need somewhere to read about the new option. It looks like aten/src/ATen/native/README.md would be the place to document this?

zou3519 committed:
Update on "[namedtensor] Add named_guard to native_functions options"

[namedtensor] Add named_guard to native_functions options

If named_guard=True (default), if any of the functions' inputs are named
tensors with non-None names, then the function errors out. The plan is
to incrementally write named inference rules and setting
named_guard=False once a function supports named tensors.

Test Plan:

Read generated code. For example:
```
Tensor & CPUType::_cat_out(Tensor & out, TensorList tensors, int64_t dim) const {
    if (out.is_named() || at::has_names(tensors)) {
        AT_ERROR("_cat_out: no named inference rule implemented.");
    }
    const OptionalDeviceGuard device_guard(device_of(out));
    return at::native::legacy::cpu::_th_cat_out(/* native_actuals */ out, tensors, dim);
}
```

gh-metadata: pytorch pytorch 21373 gh/zou3519/42/head
